dipy/reconst/mcsd.py

```diff
@@ -0,0 +1,347 @@
+
+    # Realign reg_sphere so that the first vertex is aligned with delta
+    # orientation (theta, phi).
+    M = geo.vec2vec_rotmat(reg_sphere.vertices[0], [x, y, z])
+    new_vertices = np.dot(reg_sphere.vertices, M.T)
+    _, t, p = geo.cart2sphere(*new_vertices.T)
+
+    B = shm.real_sph_harm(m, n, t[:, None], p[:, None])
+    G_temp = np.ascontiguousarray(B[:, n != 0])
+    # c_ samples the delta function at the delta orientation.
+    c_temp = G_temp[0][:, None]
+    a_temp, b_temp = G_temp.shape
+
+    c_int = cvx.Parameter((c_temp.shape[0], 1))
+    c_int.value = -c_temp
+    G = cvx.Parameter((a_temp, b_temp))
+    G.value = -G_temp
+    h = cvx.Parameter((a_temp, b_temp))
+    h_temp = np.full((a_temp, b_temp), SH_CONST ** 2)
+    h.value = h_temp
+
+    # n == 0 is set to sh_const to ensure a normalized delta function.
+    # n > 0 values are optimized so that delta > 0 on all points of the sphere
+    # and delta(theta, phi) is maximized.
+    lp_prob = cvx.Problem(cvx.Maximize(cvx.sum(c_temp)), [G*x <= h])
+    r = lp_prob.solve()
+    out = np.zeros(B.shape[1])
+    out[n == 0] = SH_CONST
+    out[n != 0] = r
+
+    iso_d = [SH_CONST] * iso
+    return np.concatenate([iso_d, out])
+
+
+delta_functions = {"basic": _basic_delta,
+                   "positivity_constrained": _pos_constrained_delta}
+
+
+class MultiShellDeconvModel(shm.SphHarmModel):
+    def __init__(self, gtab, response, reg_sphere=default_sphere, iso=2,
+                 delta_form='basic'):
+        r"""
+        Multi-Shell Multi-Tissue Constrained Spherical Deconvolution
+        (MSMT-CSD) [1]_. This method extends the CSD model proposed in [2]_ by
+        the estimation of multiple response functions as a function of multiple
+        b-values and multiple tissue types.
+
+        Spherical deconvolution computes a fiber orientation distribution
+        (FOD), also called fiber ODF (fODF) [2]_. The fODF is derived from
+        different tissue types and thus overcomes the overestimation of WM in
+        GM and CSF areas.
+
+        The response function is based on the different tissue types
+        and is provided as input to the MultiShellDeconvModel.
+        It will be used as deconvolution kernel, as described in [2]_.
+
+        Parameters
+        ----------
+        gtab : GradientTable
+        response : tuple or AxSymShResponse object
+            A tuple with two elements. The first is the eigen-values as an (3,)
+            ndarray and the second is the signal value for the response
+            function without diffusion weighting.  This is to be able to
+            generate a single fiber synthetic signal. The response function
+            will be used as deconvolution kernel ([1]_)
+        reg_sphere : Sphere (optional)
+            sphere used to build the regularization B matrix.
+            Default: 'symmetric362'.
+            weight given to the constrained-positivity regularization part of
```

Comment:
Again this text about the weight. Does it really belong together with `reg_sphere`?